README.md
# Terraform-Profider-Bigv

[BigV](http://bigv.io) is a private dedicated virtual machine platfom provided by Bytemark,

This terraform provider allows the management of machines on that platform.

See [BigV prices](http://www.bigv.io/prices), [Resource definitions](http://www.bigv.io/support/api/definitions/) and 
[API documentation](http://www.bigv.io/support/api/virtual_machines/) for configuration options to resource parameters.

## Imaging and status of created

If the os you provide is not none, then terraform will wait for the imaging to finish before reporting completion.
That should be one in under 3 minutes depending on your distribution.
Your distribution must expose an ssh daemon on port 22, since that's how we define "alive".

If your os is none, then we either wait for the VM to be processed by bigv if power_on is false,
or we wait for it to be powered up if power_on is true.

## Resource changes and reboots

If you change cores or memory, the VM *will* be restarted.
It's always necessary for increases. Strictly bigv can downsize without restarts,
but we've noticed that decreasing RAM nearly always ends up inconsistent in the VM.
e.g. decreasing to 1GiB gives you 750MiB until you restart.

## Provider parameters

* **account**

   Bigv account name.

* **user**

   Bigv Username

* **password**

   Your bigv password.
   Yubikey is not yet supported. Patches welcome.

## Resource parameters

* **name**

   The VM name

* **group**

   The group name for the server.

   Defaults to default.

* **zone**

   The zone to put the server in. Currently this is machester or york. See [definitions](http://www.bigv.io/support/api/definitions/).

   Defaults to york.

* **cores**

   How many cores to allocate. Note that cores must be allocated 1 per 4GiB of RAM.
   So 1-3GiB = 1 core, 4-7GiB = 2 core.

   If you don't specify cores, but do specify memory, you'll be assigned the correct cores for the memory.

   Defaults to 1.

* **memory**

   How much RAM to allocate in MiB.

   If you don't specify mmeory byt do specify cores, you'll be allocated the minimal memory for the core count.
   e.g. 1GiB for 1 core, 4GiB for 2 cores, etc.

   Defaults to 1024.

* **os**

   Short name of operating system to image.
   See: [Resource definitions](http://www.bigv.io/support/api/definitions/)
   Common options are: jessie, vivid and none

   Defaults to none.

* **ipv4**
* **ipv6**

   IP address to allocate.
   We recommend ips should be specified because it eases the burden on bytemark's allocation process,
   and should allow concurrent imaging without deadlocks.

* **disc_size**

   Disc size in MiB. More options in the API are not yet supported, such as storage grade.

   Defaults to 25600

* **power_on**

   Whether or not the machine should be powered.
   Note that if *reboot* is true then setting *power_on* to false just reboots the VM immediately.
   See [Bigv power documentation](http://www.bigv.io/support/api/virtual_machines/#Power)
   
   Defaults to true.

* **reboot**

   Whether or not the machine should be rebooted automatically when powered down.
   See [Bigv power documentation](http://www.bigv.io/support/api/virtual_machines/#Power)
   
   Defaults to true.

* **ssh_public_key**

   SSH public key to be created on the VM. Can be multiple keys.

* **firstboot_script**

   A script to be run on first boot only by the bigv system itself.
   Useful for provisioning, especially if terraform remote-exec or file provisioners don't work.

## Computed values

* **root_password**

   The root password assigned to this vm.

## Example Usage

variables.tf:
```
variable "account" {
  default = "myaccount"
}
variable "user" {
  default = "myuser"
}
// Probably best to supply this interactively
variable "password" { }
```

provider.tf:
```
provider "bigv" {
  account  = "${var.account}"
  user     = "${var.user}"
  password = "${var.password}"
}
```

resources.tf:
```
resource "bigv_vm" "tf01" {
  name    = "tf01"
  ipv4    = "49.48.12.201"
  ipv6    = "1996:41c8:20:5ed::3:1"
  cores   = 1
  memory  = 1024
  os      = "vivid"
  group   = "default"
  zone    = "manchester"
}
 
resource "bigv_vm" "tf02" {
  name    = "tf02"
  ipv4    = "49.48.12.202"
  ipv6    = "1996:41c8:20:5ed::3:2"
  cores   = 2
  memory  = 4096
  os      = "trusty"
  group   = "vlan1519"
  zone    = "york"
}
 
resource "bigv_vm" "tf03" {
  name    = "tf03"
  ipv4     = "49.48.12.203"
  ipv6    = "1996:41c8:20:5ed::3:3"
  cores   = 1
  memory  = 1024
  os      = "none"
} 
```

And then just:
```
terraform plan
terraform apply
terraform refresh
terraform destroy
```

## Debugging and troubleshooting

Terraform commands support TF_LOG=level environment variables:
```
TF_LOG=DEBUG terraform apply
```
This provider will be fairly verbose.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [UNRELEASED] - 0000-00-00
### Added
- Allow specifying storage_grade

## [1.4.1] - 2016-03-31
### Fixed
- Fix vm create always erroring and never creating anything

## [1.4.0] - 2016-03-31
### Added
- Allow dynamic ips from bigv
### Fixed
- Fix crash on ipv6 being set from bigv

## [1.3.2] - 2016-02-20
### Fixed
- Fix blocking when a vm create fails

## [1.3.1] - 2016-02-13
### Fixed
- Fix fresh session id never being used

## [1.3.0] - 2016-02-13
### Changed
- Remove all non-alphanumetic password characters, and increase password size to compensate
### Fixed
- Fix login retries with 401 being skipped

## [1.2.1] - 2016-02-05
### Fixed
- Fix provider not setting connection info for provisioners.

## [1.2.0] - 2016-02-03
### Added
- firstboot_script attribute for bootstrapping
### Changed
- Increasse VM provisioning timeout to 20 minutes to allow for firstboot scripts
- Ignore most ssh errors, but log them. Often we get errors as ssh comes up for the first time

## [1.1.0] - 2016-02-01
### Changed
- Synchronous creation now only applies to the initial request to vm_create, which should increase parellism
### Added
- Creation now waits for the VM to be imaged fully
- Partial state is now supported for creation
### Fixed
- Root password being lost after vm create

## [1.0.0] - 2016-01-31
### Added
- Initial full release of bigv provider


bigv/client.go
package bigv

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"sync"
	"time"
)

const bigvUri = "https://uk0.bigv.io"
const bigvAuthUri = "https://auth.bytemark.co.uk/session"
const bigvTimeout = 20

type client struct {
	account  string
	user     string
	password string
	http     *http.Client
	session  string
}

var sessions sync.Mutex

type credentials struct {
	Username string `json:"username"`
	Password string `json:"password"`
}

func (c *client) newSession() error {
	l := log.New(os.Stderr, "", 0)

	cr := credentials{
		Username: c.user,
		Password: c.password,
	}

	body, err := json.Marshal(cr)
	if err != nil {
		l.Printf("Error creating json: %s", err)
		return err
	}

	l.Printf("Requesting new session at: %s", bigvAuthUri)
	req, _ := http.NewRequest("POST", bigvAuthUri, bytes.NewBuffer(body))
	req.Header.Add("Content-Type", "application/json")
	req.Header.Add("Accept", "text/plain")

	if resp, err := c.http.Do(req); err != nil {
		return err
	} else {

		body, _ := ioutil.ReadAll(resp.Body)
		defer resp.Body.Close()

		c.session = string(body)
		l.Printf("Got back session Id: %s", c.session)
	}

	return nil
}

func (c *client) do(req *http.Request) (*http.Response, error) {
	l := log.New(os.Stderr, "", 0)

	if c.http == nil {
		// Initialization
		c.http = &http.Client{
			Timeout: time.Second * bigvTimeout,
		}
	}

	if c.session == "" {
		sessions.Lock()
		// Check again, in case it's been fixed by something we were blocking on
		if c.session == "" {
			if err := c.newSession(); err != nil {
				return nil, err
			}
		}
		sessions.Unlock()
	}

	req.Header.Set("Content-Type", "application/json")

	// We're going to potentially do this again, so we need to copy the body
	var body []byte
	if req.Body != nil {
		body, _ = ioutil.ReadAll(req.Body)
	}

	for i := 0; i < 3; i++ {
		if len(body) > 0 {
			req.Body = ioutil.NopCloser(bytes.NewReader(body))
		}

		// Set inside for loop because we regenerate it if we 401
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.session))
		l.Printf("Using Session Id: %s", c.session)

		resp, err := c.http.Do(req)

		if err != nil {
			return resp, err
		}

		// A good response
		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			return resp, err
		}

		// Otherwise we need to massage and deal with auth retries

		if resp.StatusCode == 401 && i == 0 {
			l.Printf("HTTP 401. Retrying with a new session id")
			time.Sleep(1 * time.Second)
			c.newSession()
			continue
		}

		// Any other http error. Try to get more about it
		body, _ := ioutil.ReadAll(resp.Body)
		defer resp.Body.Close()
		return resp, fmt.Errorf("Bigv returned HTTP Status %d: %s", resp.StatusCode, body)
	}

	return nil, errors.New("Unexpected error in HTTP client, this should not happen")
}


bigv/provider.go
package bigv

import (
	"github.com/hashicorp/terraform/helper/schema"
	"github.com/hashicorp/terraform/terraform"
)

func Provider() terraform.ResourceProvider {
	return &schema.Provider{
		Schema: map[string]*schema.Schema{
			"account": &schema.Schema{
				Type:        schema.TypeString,
				Required:    true,
				DefaultFunc: schema.EnvDefaultFunc("BIGV_ACCOUNT", nil),
				Description: "The bigv account name",
			},
			"user": &schema.Schema{
				Type:        schema.TypeString,
				Required:    true,
				DefaultFunc: schema.EnvDefaultFunc("BIGV_USER", nil),
				Description: "The bigv user name",
			},
			"password": &schema.Schema{
				Type:        schema.TypeString,
				Required:    true,
				DefaultFunc: schema.EnvDefaultFunc("BGIV_PASSWORD", nil),
				Description: "The bigv password",
			},
		},
		ResourcesMap: map[string]*schema.Resource{
			"bigv_vm": resourceBigvVM(),
		},
		ConfigureFunc: providerConfigure,
	}
}

func providerConfigure(d *schema.ResourceData) (bigvClient interface{}, err error) {

	bigvClient = &client{
		account:  d.Get("account").(string),
		user:     d.Get("user").(string),
		password: d.Get("password").(string),
	}

	return
}


bigv/resource_vm.go
package bigv

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"math/rand"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/hashicorp/terraform/helper/schema"
	"golang.org/x/crypto/ssh"
)

const (
	passwordLength     = 48
	waitForVM          = 1200
	vmCheckInterval    = 5
	waitForProvisioned = 1 + iota
	waitForPowered     = 1 + iota
)

type bigvVm struct {
	Id           int    `json:"id,omitempty"`
	Name         string `json:"name,omitempty"`
	Cores        int    `json:"cores,omitempty"`
	Memory       int    `json:"memory,omitempty"`
	Hostname     string `json:"hostname,omitempty"`
	Distribution string `json:"last_imaged_with,omitempty"`
	Power        bool   `json:"power_on"`
	Reboot       bool   `json:"autoreboot_on"`
	Group        string `json:"group,omitempty"`
	GroupId      int    `json:"group_id,omitempty"`
	Zone         string `json:"zone_name,omitempty"`
}

type bigvDisc struct {
	Label        string `json:"label,omitempty"`
	StorageGrade string `json:"storage_grade,omitempty"`
	Size         int    `json:"size,omitempty"`
}

type bigvImage struct {
	Distribution    string `json:"distribution,omitempty"`
	RootPassword    string `json:"root_password,omitempty"`
	SshPublicKey    string `json:"ssh_public_key,omitempty"`
	FirstBootScript string `json:"firstboot_script,omitempty"`
}

type bigvIps struct {
	// Create attributes
	Ipv4 string `json:"ipv4,omitempty"`
	Ipv6 string `json:"ipv6,omitempty"`
}

type bigvNic struct {
	// Read Attributes
	Label string   `json:"label,omitempty"`
	Ips   []string `json:"ips,omitempty"`
	Mac   string   `json:"mac,omitempty"`
}

type bigvServer struct {
	bigvVm
	Discs []bigvDisc `json:"discs,omitempty"`
	Nics  []bigvNic  `json:"network_interfaces,omitempty"`
}

type bigvVMCreate struct {
	VirtualMachine bigvVm     `json:"virtual_machine"`
	Discs          []bigvDisc `json:"discs,omitempty"`
	Image          bigvImage  `json:"reimage,omitempty"`
	Ips            *bigvIps   `json:"ips,omitempty"` // Just used for create
}

func resourceBigvVM() *schema.Resource {
	return &schema.Resource{
		Create: resourceBigvVMCreate,
		Read:   resourceBigvVMRead,
		Update: resourceBigvVMUpdate,
		Delete: resourceBigvVMDelete,
		Exists: resourceBigvVMExists,
		Schema: map[string]*schema.Schema{
			"name": &schema.Schema{
				Type:     schema.TypeString,
				Required: true,
				ForceNew: true,
			},
			"group": &schema.Schema{
				Type:        schema.TypeString,
				Optional:    true,
				ForceNew:    true,
				Default:     "default",
				Description: "bigv group name for the VM. Defaults to default",
			},
			"group_id": &schema.Schema{
				Type:     schema.TypeInt,
				Computed: true,
			},
			"zone": &schema.Schema{
				Type:        schema.TypeString,
				Optional:    true,
				ForceNew:    true,
				Default:     "york",
				Description: "bigv zone to put the VM in. Defaults to york",
			},
			"ipv4": &schema.Schema{
				Type:     schema.TypeString,
				Optional: true,
				Computed: true,
			},
			"ipv6": &schema.Schema{
				Type:     schema.TypeString,
				Optional: true,
				Computed: true,
			},
			"os": &schema.Schema{
				Type:     schema.TypeString,
				Optional: true,
				Default:  "vivid",
				ForceNew: true,
			},
			"cores": &schema.Schema{
				Type:         schema.TypeInt,
				Optional:     true,
				Computed:     true,
				ComputedWhen: []string{"memory"},
			},
			"memory": &schema.Schema{
				Type:         schema.TypeInt,
				Optional:     true,
				Computed:     true,
				ComputedWhen: []string{"cores"},
			},
			"disc_size": &schema.Schema{
				Type:     schema.TypeInt,
				Optional: true,
				Default:  "25600",
				ForceNew: true,
			},
			"storage_grade": &schema.Schema{
				Type:     schema.TypeString,
				Optional: true,
				Default:  "sata",
				ForceNew: true,
			},
			"root_password": &schema.Schema{
				Type:     schema.TypeString,
				Computed: true,
			},
			"power_on": &schema.Schema{
				Type:        schema.TypeBool,
				Default:     true,
				Optional:    true,
				Description: "Whether or not the VM should be powered. Note that with reboot true, power_on false is just a reboot",
			},
			"reboot": &schema.Schema{
				Type:        schema.TypeBool,
				Default:     true,
				Optional:    true,
				Description: "Whether or not to reboot the VM when the power_on is turned off",
			},
			"ssh_public_key": &schema.Schema{
				Type:        schema.TypeString,
				Optional:    true,
				Description: "One or more ssh public keys to put on the machine. Will only work if os is not core",
			},
			"firstboot_script": &schema.Schema{
				Type:        schema.TypeString,
				Optional:    true,
				Description: "A script to be executed on first boot arbitrarily",
			},
		},
	}
}

var createPipeline sync.Mutex

func resourceBigvVMCreate(d *schema.ResourceData, meta interface{}) error {

	l := log.New(os.Stderr, "", 0)

	bigvClient := meta.(*client)

	vm := bigvVMCreate{
		VirtualMachine: bigvVm{
			Name:   d.Get("name").(string),
			Cores:  d.Get("cores").(int),
			Memory: d.Get("memory").(int),
			Power:  d.Get("power_on").(bool),
			Reboot: d.Get("reboot").(bool),
			Group:  d.Get("group").(string),
			Zone:   d.Get("zone").(string),
		},
		Discs: []bigvDisc{{
			Label:        "root",
			StorageGrade: d.Get("storage_grade").(string),
			Size:         d.Get("disc_size").(int),
		}},
		Image: bigvImage{
			Distribution:    d.Get("os").(string),
			RootPassword:    randomPassword(),
			SshPublicKey:    d.Get("ssh_public_key").(string),
			FirstBootScript: d.Get("firstboot_script").(string),
		},
	}

	// If no ipv* is set then let bigv allocate it itself
	// The json for ip must be nil
	if ip := d.Get("ipv4"); ip != nil && ip.(string) != "" {
		vm.Ips = &bigvIps{
			Ipv4: ip.(string),
		}
	}

	if ip := d.Get("ipv6"); ip != nil && ip.(string) != "" {
		if vm.Ips == nil {
			vm.Ips = &bigvIps{}
		}
		vm.Ips.Ipv6 = ip.(string)
	}

	// Make sure the root password gets stored in d
	d.Set("root_password", vm.Image.RootPassword)

	// Connection information
	connInfo := map[string]string{
		"type":     "ssh",
		"password": vm.Image.RootPassword,
	}
	if vm.Ips != nil {
		connInfo["host"] = vm.Ips.Ipv4
	}
	d.SetConnInfo(connInfo)

	if err := vm.VirtualMachine.computeCoresToMemory(); err != nil {
		return err
	}

	if vm.Image.SshPublicKey != "" && vm.Image.Distribution == "none" {
		return errors.New("Cannot deploy ssh public keys with an os of 'none'. Please use a provisioner instead")
	}

	body, err := json.Marshal(vm)
	if err != nil {
		return err
	}

	// VM create uses a bigger path
	url := fmt.Sprintf("%s/accounts/%s/groups/%s/vm_create",
		bigvUri,
		bigvClient.account,
		vm.VirtualMachine.Group, // this will be group name
	)

	l.Printf("Requesting VM create: %s", url)
	l.Printf("VM profile: %s", body)

	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(body))

	// TODO - Early 2016, and we hope to remove this soonish
	// bigV deadlocks if you hit it with concurrent creates.
	// That might be an ip allocation issue, and specifying both ips might
	// fix it, but that's untested. For now waiting for them to confirm we
	// can lift this restriction.
	createPipeline.Lock()
	resp, err := bigvClient.do(req)
	createPipeline.Unlock()
	if err != nil {
		return err
	}

	// Always close the body when done
	defer resp.Body.Close()

	l.Printf("HTTP response Status: %s", resp.Status)

	if resp.StatusCode != http.StatusAccepted {
		body, _ := ioutil.ReadAll(resp.Body)
		return fmt.Errorf("Create VM status %d from bigv: %s", resp.StatusCode, body)
	}

	d.Partial(true)
	for _, i := range []string{"name", "group_id", "group", "zone", "cores", "memory", "ipv4", "ipv6", "root_password"} {
		d.SetPartial(i)
	}

	for k, v := range resp.Header {
		l.Printf("%s: %s", k, v)
	}

	// wait for state also sets up the resource from the read state we get back
	if err := waitForBigvState(d, bigvClient, waitForProvisioned); err != nil {
		return err
	}

	l.Printf("Created BigV VM, Id: %s", d.Id())

	// If we expect it to be turned on, wait for it to powered
	if vm.VirtualMachine.Power == true {
		if err := waitForBigvState(d, bigvClient, waitForPowered); err != nil {
			return err
		}

		// This assumes all distributions will listen on public ssh
		if vm.Image.Distribution != "none" {
			if err := waitForVmSsh(d); err != nil {
				return err
			}
		}
	}

	d.Partial(false)

	return nil

}

// waitForBigvState
// Obviously wait for a state
// Also sets up the resource from the state read
func waitForBigvState(d *schema.ResourceData, bigvClient *client, waitFor int) error {
	l := log.New(os.Stderr, "", 0)

	url := fmt.Sprintf("%s/virtual_machines/%s?view=overview",
		bigvUri,
		d.Get("name"),
	)

	l.Printf("VM Health Check: %s", url)
	req, _ := http.NewRequest("GET", url, nil)

	var body []byte
	for {
		select {
		case <-time.After(waitForVM * time.Second):
			return fmt.Errorf("VM state didn't happen in %d seconds", waitForVM)
		case <-time.Tick(vmCheckInterval * time.Second):
			resp, err := bigvClient.do(req)
			if err != nil {
				return fmt.Errorf("Error checking on VM health: %s", err)
			}

			// Always close the body when done
			defer resp.Body.Close()

			body, _ = ioutil.ReadAll(resp.Body)

			l.Printf("HTTP response Status: %s", resp.Status)
			// No matter what, update everything comes from the state
			if err := resourceFromJson(d, body); err != nil {
				return err
			}

			if resp.StatusCode == http.StatusOK {
				if waitFor == waitForProvisioned {
					l.Println("VM is Up and HTTP OK")
					return nil
				}

				l.Println("VM power:", d.Get("power_on").(bool))
				switch {
				case waitFor == waitForPowered && d.Get("power_on").(bool):
					l.Println("VM is powered")
					return nil
				}
			}

			if resp.StatusCode != http.StatusAccepted && resp.StatusCode != http.StatusOK {
				return fmt.Errorf("VM healthcheck Bad HTTP status from bigv: %d", resp.StatusCode)
			}
		}
	}

	return fmt.Errorf("VM state didn't happen in %d seconds", waitForVM)
}

// Simply waits for ssh to come up
func waitForVmSsh(d *schema.ResourceData) error {
	l := log.New(os.Stderr, "", 0)

	l.Printf("Waiting for VM ssh: %s", d.Get("name"))

	config := &ssh.ClientConfig{
		User: "root",
		Auth: []ssh.AuthMethod{
			ssh.Password(d.Get("root_password").(string)),
		},
	}

	for {
		select {
		case <-time.After(waitForVM * time.Second):
			return fmt.Errorf("VM ssh wasn't up in %d seconds", waitForVM)
		case <-time.Tick(vmCheckInterval * time.Second):
			conn, err := ssh.Dial("tcp", fmt.Sprintf("%s:22", d.Get("ipv4")), config)
			if err != nil {
				if strings.Contains(err.Error(), "connection refused") {
					l.Println("SSH isn't up yet")
					continue
				} else {
					l.Printf("SSH Error, ignored: %s", err.Error())
					continue
				}
			}
			conn.Close()
			l.Println("SSH alive and kicking")
			return nil
		}
	}

	return errors.New("Ssh wait should never get here")
}

func resourceBigvVMUpdate(d *schema.ResourceData, meta interface{}) error {
	l := log.New(os.Stderr, "", 0)

	bigvClient := meta.(*client)

	vm := bigvVm{}

	if d.HasChange("power_on") {
		vm.Power = d.Get("power_on").(bool)
	}

	if d.HasChange("power_on") {
		vm.Reboot = d.Get("reboot").(bool)
	}

	if d.HasChange("cores") || d.HasChange("memory") {
		// Specifiy both cores and memory together always, so we can validate them.
		vm.Cores = d.Get("cores").(int)
		vm.Memory = d.Get("memory").(int)

		// Whenever we change either of these reboot the server
		// That's because even though decreasing ram doesn't require a reboot,
		// it looks like it often goes wrong and you get less ram than you should.
		// e.g. lowering to 1GB nearly always gives you 750MB
		if !d.HasChange("power_on") {
			vm.Power = false
			// Always need Reboot on, otherwise it'll stay down
			vm.Reboot = true
		}
	}

	if err := vm.computeCoresToMemory(); err != nil {
		return err
	}

	body, err := json.Marshal(vm)
	if err != nil {
		return err
	}

	url := fmt.Sprintf("%s/accounts/%s/groups/%s/virtual_machines/%s",
		bigvUri,
		bigvClient.account,
		d.Get("group"),
		d.Id(),
	)

	l.Printf("Requesting VM update: %s", url)
	l.Printf("VM profile: %s", body)

	req, err := http.NewRequest("PUT", url, bytes.NewBuffer(body))
	if err != nil {
		l.Printf("Error creating request for Update: %s", err)
		return err
	}

	if resp, err := bigvClient.do(req); err != nil {
		return err
	} else {

		// Always close the body when done
		defer resp.Body.Close()

		l.Printf("HTTP response Status: %s", resp.Status)

		if resp.StatusCode != http.StatusOK {
			return fmt.Errorf("Update VM bad status from bigv: %d", resp.StatusCode)
		}

		if body, err := ioutil.ReadAll(resp.Body); err != nil {
			return err
		} else {
			if err := resourceFromJson(d, body); err != nil {
				return err
			}
		}

		l.Printf("Updated BigV VM, Id: %s", d.Id())

		return nil
	}
	return nil
}

func resourceBigvVMRead(d *schema.ResourceData, meta interface{}) error {
	l := log.New(os.Stderr, "", 0)

	bigvClient := meta.(*client)

	url := fmt.Sprintf("%s/virtual_machines/%s?view=overview",
		bigvUri,
		d.Get("name"),
	)

	l.Printf("VM Read: %s", url)

	req, _ := http.NewRequest("GET", url, nil)

	resp, err := bigvClient.do(req)
	if err != nil {
		return err
	}

	// Always close the body when done
	defer resp.Body.Close()

	l.Printf("HTTP response Status: %s", resp.Status)

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("Read VM Bad HTTP status from bigv: %d", resp.StatusCode)
	}

	body, ioErr := ioutil.ReadAll(resp.Body)
	if ioErr != nil {
		return ioErr
	}

	return resourceFromJson(d, body)
}

func resourceBigvVMDelete(d *schema.ResourceData, meta interface{}) error {
	l := log.New(os.Stderr, "", 0)

	bigvClient := meta.(*client)

	url := fmt.Sprintf("%s/accounts/%s/groups/%s/virtual_machines/%s?purge=true",
		bigvUri,
		bigvClient.account,
		d.Get("group"),
		d.Id(),
	)
	l.Printf("Deleting VM at %s", url)
	req, err := http.NewRequest("DELETE", url, nil)
	if err != nil {
		return err
	}

	if resp, err := bigvClient.do(req); err != nil {
		return err
	} else {
		// Always close the body when done
		defer resp.Body.Close()

		l.Printf("Delete %s HTTP response Status: %s", d.Id(), resp.Status)
		if resp.StatusCode != http.StatusNoContent {
			return fmt.Errorf("Delete VM %s Bad HTTP status from bigv: %d", d.Id(), resp.StatusCode)
		}
	}

	return nil
}

func resourceBigvVMExists(d *schema.ResourceData, meta interface{}) (bool, error) {
	l := log.New(os.Stderr, "", 0)

	bigvClient := meta.(*client)

	url := fmt.Sprintf("%s/virtual_machines/%s",
		bigvUri,
		d.Id(),
	)

	l.Printf("Checking VM existance at %s", url)

	req, _ := http.NewRequest("GET", url, nil)
	resp, err := bigvClient.do(req)
	if err != nil {
		return false, err
	}

	l.Printf("Exists %s HTTP response Status: %s", d.Id(), resp.Status)
	if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusAccepted {
		return true, nil
	} else if resp.StatusCode == http.StatusNotFound {
		return false, nil
	}

	return false, fmt.Errorf("Unexpected HTTP status from VM exists check: %d", resp.StatusCode)
}

func resourceFromJson(d *schema.ResourceData, vmJson []byte) error {
	l := log.New(os.Stderr, "", 0)

	l.Printf("VM definition: %s", vmJson)

	vm := &bigvServer{}

	if err := json.Unmarshal(vmJson, vm); err != nil {
		return err
	}

	d.SetId(strconv.Itoa(vm.Id))
	d.Set("name", vm.Name)
	d.Set("cores", vm.Cores)
	d.Set("memory", vm.Memory)
	d.Set("power_on", vm.Power)
	d.Set("reboot", vm.Reboot)
	d.Set("group_id", vm.GroupId)
	d.Set("zone", vm.Zone)

	// If we don't get discs back, this was probably an update request
	if len(vm.Discs) == 1 {
		d.Set("disk_size", vm.Discs[0].Size)
	}

	// Distribution is empty in create response, leave it with what we sent in
	if vm.Distribution != "" {
		d.Set("os", vm.Distribution)
	}

	// Not finding the ips is fine, because they're not sent back in the create request
	if len(vm.Nics) > 0 {
		// This is fairly^Wvery^Wacceptably hacky
		d.Set("ipv4", vm.Nics[0].Ips[0])
		d.Set("ipv6", vm.Nics[0].Ips[1])

		d.SetConnInfo(map[string]string{
			"type":     "ssh",
			"host":     vm.Nics[0].Ips[0],
			"password": d.Get("root_password").(string),
		})
	}

	return nil
}

/* computeCoresToMemory
bigv charges per 1GiB memory, and you automatically get 1 more core per 4GiB.
See: http://www.bigv.io/prices
*/
func (v *bigvVm) computeCoresToMemory() error {

	switch {
	case v.Cores == 0 && v.Memory == 0:
		// Both unset, so just supply defaults
		v.Cores = 1
		v.Memory = 1024
	case v.Cores == 0:
		// Just the cores calculated
		v.Cores = int(math.Ceil(float64(v.Memory/4096) + float64(0.01)))
	case v.Memory == 0:
		// Just the memory calculated
		v.Memory = int(math.Max(1024, float64((v.Cores-1)*4096)))
	default:
		// Both set, so validate them
		expectedCores := int(math.Ceil(float64(v.Memory/4096) + float64(0.01)))
		if expectedCores != v.Cores {
			return fmt.Errorf("Memory and cores mismatch!\nExpected %d cores for your %dGiB memory, but you have %d.\nSpecify 1 cores per 4GiB memory.\nSee: http://www.bigv.io/prices", expectedCores, v.Memory/1024, v.Cores)
		}
	}

	return nil
}

const letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"

func randomPassword() string {
	r := rand.New(rand.NewSource(time.Now().UnixNano()))

	b := make([]byte, passwordLength)
	for i := range b {
		b[i] = letters[r.Intn(len(letters))]
	}
	return string(b)
}
